src/lib.rs

```diff
@@ -52,6 +51,58 @@ fn repeat_byte(b: u8) -> usize {
     rep
 }
 
+/// An iterator for memchr
+pub struct Memchr<'a> {
+    needle: u8,
+    // The haystack to iterate over
+    haystack: &'a [u8],
+    // The index
+    position: usize,
+}
+
+impl<'a> Memchr<'a> {
+    /// Create a new Memchr that's initalized to zero with a haystack
+    pub fn new(needle: u8, haystack: &[u8]) -> Memchr {
+        Memchr {
+            needle: needle,
+            haystack: haystack,
+            position: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for Memchr<'a> {
+    type Item = usize;
+
+    fn next(&mut self) -> Option<usize> {
+        let search_result = memchr(self.needle, &self.haystack);
+        match search_result {
+            Some(index) => {
+                // Move our internal position
+                self.haystack = self.haystack.split_at(index + 1).1;
+                self.position = self.position + index + 1;
+                Some(self.position)
+            }
+            None => None,
+        }
+    }
+}
+
+impl<'a> DoubleEndedIterator for Memchr<'a> {
+    fn next_back(&mut self) -> Option<Self::Item> {
+        let search_result = memrchr(self.needle, &self.haystack);
+        match search_result {
+            Some(index) => {
+                // Move our internal position
+                self.haystack = self.haystack.split_at(index).0;
+                self.position = self.position + index + 1;
+                Some(self.position)
```

Comment:
"position" is the front position, and it shouldn't move in next_back. The iterator element should be `position + index` here, shouldn't it? (A.k.a postion + haystack.len() after haystack is updated).

If that doesn't seem right, add more quickchecks. :D Exhaustive testing is better than single tests. (So yay quickcheck).